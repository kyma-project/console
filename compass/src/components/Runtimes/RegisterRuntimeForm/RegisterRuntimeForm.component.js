import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import './RegisterRuntimeForm.scss';
import { InlineHelp } from 'fundamental-react/InlineHelp';
import { CustomPropTypes } from 'react-shared';

const RegisterRuntimeForm = ({
  formElementRef,
  onChange,
  onCompleted,
  onError,
  addRuntime,
}) => {
  const formValues = {
    name: useRef(null),
    description: useRef(null),
  };

  const handleFormSubmit = async e => {
    e.preventDefault();
    const runtimeName = formValues.name.current.value;
    try {
      await addRuntime({
        name: runtimeName,
        description: formValues.description.current.value,
      });
      onCompleted(runtimeName, `Runtime created succesfully`);
    } catch (e) {
      onError(`The runtime could not be created succesfully`, e.message || ``);
    }
  };

  const nameField = () => (
    <>
      <label className="fd-form__label" htmlFor="runtime-name">
        Name *
        <InlineHelp
          placement="bottom-right"
          text="Name must be no longer than 63 characters, must start and end with a lowercase letter or number, and may contain lowercase letters, numbers, and dashes."
        />
      </label>
      <input
        className="fd-form__control"
        ref={formValues.name}
        type="text"
        id="runtime-name"
        placeholder="Runtime name"
        aria-required="true"
        required
        pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
      />
    </>
  );

  return (
    <form
      onChange={onChange}
      ref={formElementRef}
      onSubmit={handleFormSubmit}
      className="register-runtime-form"
    >
      <div className="fd-form__set">
        <div className="fd-form__item">{nameField()}</div>
        <div className="fd-form__item">
          <label className="fd-form__label" htmlFor="runtime-desc">
            Description
          </label>

          <input
            className="fd-form__control"
            ref={formValues.description}
            type="text"
            id="runtime-desc"
            placeholder="Runtime description"
          />
        </div>
      </div>
    </form>
  );
};
RegisterRuntimeForm.propTypes = {
  formElementRef: CustomPropTypes.ref, // used to store <form> element reference
  isValid: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired, // args: title(string), message(string)
  onCompleted: PropTypes.func.isRequired, // args: title(string), message(string)
};

export default RegisterRuntimeForm;
