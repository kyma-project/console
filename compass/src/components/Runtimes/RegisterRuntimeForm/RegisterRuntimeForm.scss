.fd-form__control:not(:placeholder-shown):invalid {
  border: 1px solid var(--fd-color-status-3);
  animation: gentleRedBorderPulse 2s ease infinite;
}

@keyframes gentleRedBorderPulse {
  $color1: #d04343;
  $color2: darken(#d04343, 18%);
  0% {
    border-color: $color1;
  }
  50% {
    border-color: $color2;
  }
  100% {
    border-color: $color1;
  }
}
.fd-inline-help {
  margin-left: 3px;
}

.register-runtime-form {
  .fd-inline-help__content {
    width: 300px;
    min-width: unset;
  }
}
