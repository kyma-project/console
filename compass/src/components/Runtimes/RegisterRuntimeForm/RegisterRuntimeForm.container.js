import { graphql } from 'react-apollo';
import { compose } from 'recompose';

import { REGISTER_RUNTIME } from '../gql';

import RegisterRuntimeForm from './RegisterRuntimeForm.component';

export default compose(
  graphql(REGISTER_RUNTIME, {
    props: ({ mutate, error }) => ({
      addRuntime: data =>
        mutate({
          variables: {
            in: data,
          },
        }),
    }),
  }),
)(RegisterRuntimeForm);
