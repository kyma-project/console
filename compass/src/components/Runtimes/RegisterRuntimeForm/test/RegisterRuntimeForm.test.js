import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/react-testing';

import RegisterRuntimeForm from '../RegisterRuntimeForm.container';

describe('RegisterRuntimeForm', () => {
  // for "Warning: componentWillReceiveProps has been renamed"
  console.warn = jest.fn();

  const emptyFn = () => {};
  const emptyRef = { current: null };

  it('Renders with minimal props', () => {
    const component = renderer.create(
      <MockedProvider addTypename={false}>
        <RegisterRuntimeForm
          formElementRef={emptyRef}
          isValid={false}
          onChange={emptyFn}
          onCompleted={emptyFn}
          onError={emptyFn}
        />
      </MockedProvider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(console.warn).not.toHaveBeenCalled();
  });
});
