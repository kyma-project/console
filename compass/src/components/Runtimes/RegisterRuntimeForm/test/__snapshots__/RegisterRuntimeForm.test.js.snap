// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterRuntimeForm Renders with minimal props 1`] = `
<form
  className="register-runtime-form"
  onChange={[Function]}
  onSubmit={[Function]}
>
  <div
    className="fd-form__set"
  >
    <div
      className="fd-form__item"
    >
      <label
        className="fd-form__label"
        htmlFor="runtime-name"
      >
        Name *
        <span
          className="fd-inline-help"
        >
          <span
            className="fd-inline-help__content fd-inline-help__content--bottom-right"
          >
            Name must be no longer than 63 characters, must start and end with a lowercase letter or number, and may contain lowercase letters, numbers, and dashes.
          </span>
        </span>
      </label>
      <input
        aria-required="true"
        className="fd-form__control"
        id="runtime-name"
        pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
        placeholder="Runtime name"
        required={true}
        type="text"
      />
    </div>
    <div
      className="fd-form__item"
    >
      <label
        className="fd-form__label"
        htmlFor="runtime-desc"
      >
        Description
      </label>
      <input
        className="fd-form__control"
        id="runtime-desc"
        placeholder="Runtime description"
        type="text"
      />
    </div>
  </div>
</form>
`;
